search-provider/eks-search-app.c

```diff
@@ -0,0 +1,110 @@
+/* Copyright 2016 Endless Mobile, Inc. */
+
+#include "eks-search-app.h"
+
+#include "eks-search-provider.h"
+#include "eks-search-provider-dbus.h"
+
+#include <eos-knowledge-content.h>
+#include <systemd/sd-bus.h>
+
+/**
+ * EksSearchProvider:
+ *
+ * A search provider for a single knowledge app, to be used through dbus by the
+ * shell's global search. Requires the app id of the knowledge app it should run
+ * searches for.
+ *
+ * This search provider will activate the actual knowledge app over dbus with a
+ * result or search to display.
+ */
+struct _EksSearchApp
+{
+  GApplication parent_instance;
+
+  EkncSubtreeDispatcher *dispatcher;
+  // Hash table with app id string keys, EksSearchProvider values
+  GHashTable *app_search_providers;
+};
+
+G_DEFINE_TYPE (EksSearchApp,
+               eks_search_app,
+               G_TYPE_APPLICATION)
+
+static void
+eks_search_app_finalize (GObject *object)
+{
+  EksSearchApp *self = EKS_SEARCH_APP (object);
+
+  g_clear_object (&self->dispatcher);
+  g_clear_pointer (&self->app_search_providers, g_hash_table_unref);
+
+  G_OBJECT_CLASS (eks_search_app_parent_class)->finalize (object);
+}
+
+static gboolean
+eks_search_app_register (GApplication *application,
+                         GDBusConnection *connection,
+                         const gchar *object_path,
+                         GError **error)
+{
+  EksSearchApp *self = EKS_SEARCH_APP (application);
+
+  eknc_subtree_dispatcher_register (self->dispatcher, connection, object_path);
+  return TRUE;
+}
+
+static void
+eks_search_app_unregister (GApplication *application,
+                           GDBusConnection *connection,
+                           const gchar *object_path)
+{
+  EksSearchApp *self = EKS_SEARCH_APP (application);
+
+  eknc_subtree_dispatcher_unregister (self->dispatcher);
+}
+
+static void
+eks_search_app_class_init (EksSearchAppClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+  GApplicationClass *application_class = G_APPLICATION_CLASS (klass);
+
+  object_class->finalize = eks_search_app_finalize;
+
+  application_class->dbus_register = eks_search_app_register;
+  application_class->dbus_unregister = eks_search_app_unregister;
+}
+
+static GDBusInterfaceSkeleton *
+dispatch_subtree (EkncSubtreeDispatcher *dispatcher,
+                  gchar *subnode,
+                  EksSearchApp *self)
+{
+  EksSearchProvider *provider = g_hash_table_lookup (self->app_search_providers, subnode);
+  if (provider == NULL)
+    {
+      g_autofree gchar *app_id = NULL;
+      g_autofree gchar *prefixed = g_strconcat ("/", subnode, NULL);
+      sd_bus_path_decode (prefixed, "/", &app_id);
```

Comment:
Thought it was nice we could just rely on the actual function instead of duplicating functionality (unlike in js). But could copy this in https://github.com/systemd/systemd/blob/dd8352659c9428b196706d04399eec106a8917ed/src/basic/bus-label.c#L27
https://github.com/systemd/systemd/blob/dd8352659c9428b196706d04399eec106a8917ed/src/basic/hexdecoct.c#L54

Is that your vote?